MoreLinq/Duplicates.cs

```diff
@@ -0,0 +1,69 @@
+        /// If null, the default equality comparer for <c>TSource</c> is used.</param>
+        /// <typeparam name="T">The type of the elements in the source sequence</typeparam>
+        /// <returns>all elements of the source sequence that are duplicated, based on the provided equality comparer</returns>
+        public static IEnumerable<T> Duplicates<T>(this IEnumerable<T> source, IEqualityComparer<T>? comparer)
+            => Duplicates(source, IdFn, comparer);
+
+        /// <summary>
+        ///   Returns all duplicated elements of the given source, using the specified element equality comparer
+        /// </summary>
+        /// <param name="source">source sequence.</param>
+        /// <param name="keySelector">Projection for determining duplication</param>
+        /// <typeparam name="TSource">Type of the source sequence</typeparam>
+        /// <typeparam name="TKey">Type of the projected element</typeparam>
+        /// <returns>all elements of the source sequence that are duplicated, based on the provided equality comparer</returns>
+        public static IEnumerable<TSource> Duplicates<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector)
+            => Duplicates(source, keySelector, EqualityComparer<TKey>.Default);
+
+        /// <summary>
+        ///   Returns all duplicated elements of the given source, using the specified element equality comparer
+        /// </summary>
+        /// <param name="source">source sequence.</param>
+        /// <param name="keySelector">Projection for determining duplication</param>
+        /// <param name="comparer">The equality comparer to use to determine whether or not keys are equal.
+        /// If null, the default equality comparer for <c>TSource</c> is used.</param>
+        /// <typeparam name="TSource">Type of the source sequence</typeparam>
+        /// <typeparam name="TKey">Type of the projected element</typeparam>
+        /// <returns>all elements of the source sequence that are duplicated, based on the provided equality comparer</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="source"/> is null.</exception>
+        /// <exception cref="ArgumentNullException"><paramref name="keySelector"/> is null.</exception>
+        public static IEnumerable<TSource> Duplicates<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? comparer)
+        {
+            if (source is null) throw new ArgumentNullException(nameof(source));
+            if (keySelector is null) throw new ArgumentNullException(nameof(keySelector));
+
+            return GetDuplicates();
+
+            IEnumerable<TSource> GetDuplicates()
+            {
+                var enumeratedElements = new HashSet<TKey>(comparer);
+                foreach (var element in source)
+                {
+                    if (enumeratedElements.Add(keySelector(element)) is false)
```

Comment:
Consider the following sequence: [0, 0, 0].

1. `.Add()` returns `true` as it adds `0` to the set
2. `.Add()` returns `false` as it does not add `0` to the set.
   * `yield return` returns a `0`.
3. `.Add()` returns `false` as it does not add `0` to the set.
   * `yield return` returns a `0`.

Expected result sequence: [0]
Actual result sequence: [0, 0]

Thoughts?

Reply:
Is this by design? should we return all exceeding elements or just one of each duplicate?

Reply:
That's  very relevant observation @viceroypenguin  

it's either returning all or only one of each duplicate.

i think its better to return all since we have a KeySelector, so maybe they share a inner property value but the elements are different like:

var a = DummyClass("A", 1);
var b = DummyClass("A", 2);
var c = DummyClass("B", 3);

Given a list of (a,c,b).duplicates(x => x.StringProperty ) would return a and b
Given a list of (a,c,b).duplicates(x => x.IntProperty) would return an empty list

so i think i'll fix and add a test for the one raised @viceroypenguin  and i'll keep the behaviour as Returning ALL

what do you think? @atifaziz  @leandromoh  @viceroypenguin

Reply:
In general, at least in considering simple types, only one of each duplicate would be expected. This would match the behavior of `.Distinct()`/`.DistinctBy()`, which only returns one of each distinct value. If non-key values of complex types are important, they aren't really non-key values, so they should be included in the `keySelector`.

Reply:
Done!